src/math.cpp

```diff
@@ -355,83 +527,144 @@ void randomseed(const CallContext& ctx) {
 }
 
 auto sin(const CallContext& ctx) -> Value {
-    return math_helper(ctx, static_cast<double (*)(double)>(&std::sin), "sin");
+    auto origin = Origin(UnaryOrigin{
+        .val = make_owning<Value>(ctx.arguments().get(0)),
+        .location = ctx.call_location(),
+        .reverse = [](const Value& new_value,
+                      const Value& old_value) -> std::optional<SourceChangeTree> {
+            Number n = std::get<Number>(new_value);
+            return old_value.force(std::asin(n.value));
+        }});
+    return math_helper(ctx, static_cast<double (*)(double)>(&std::sin), "sin").with_origin(origin);
 }
 
 auto sqrt(const CallContext& ctx) -> Value {
-    return math_helper(ctx, static_cast<double (*)(double)>(&std::sqrt), "sqrt");
+    auto origin = Origin(UnaryOrigin{
+        .val = make_owning<Value>(ctx.arguments().get(0)),
+        .location = ctx.call_location(),
+        .reverse = [](const Value& new_value,
+                      const Value& old_value) -> std::optional<SourceChangeTree> {
+            Number n = std::get<Number>(new_value);
+            return old_value.force(n.value * n.value);
```

Comment:
Return `nullopt` if we try to force the value to a negative number (because lua does not support imaginary numbers).